pkg/disttask/framework/taskexecutor/task_executor.go

```diff
@@ -230,181 +251,166 @@ func (e *BaseTaskExecutor) Run(resource *proto.StepResource) {
-	), "execute task step")
-	// log as info level, subtask might be cancelled, let caller check it.
-	defer func() {
-		stepLogger.End(zap.InfoLevel, resErr)
-	}()
+func (e *BaseTaskExecutor) createStepExecutor(resource *proto.StepResource) error {
+	task := e.task.Load()
 
 	stepExecutor, err := e.GetStepExecutor(task)
 	if err != nil {
 		e.logger.Info("failed to get step executor", zap.Error(err))
-		e.failOneSubtask(runStepCtx, task.ID, err)
+		e.failOneSubtask(e.ctx, task.ID, err)
 		return errors.Trace(err)
 	}
-	execute.SetFrameworkInfo(stepExecutor, resource)
+	execute.SetFrameworkInfo(stepExecutor, task.Step, resource)
 
-	if err := stepExecutor.Init(runStepCtx); err != nil {
+	if err := stepExecutor.Init(e.ctx); err != nil {
 		if e.IsRetryableError(err) {
 			e.logger.Info("meet retryable err when init step executor", zap.Error(err))
 		} else {
 			e.logger.Info("failed to init step executor", zap.Error(err))
-			e.failOneSubtask(runStepCtx, task.ID, err)
+			e.failOneSubtask(e.ctx, task.ID, err)
 		}
 		return errors.Trace(err)
 	}
 
-	defer func() {
-		err := stepExecutor.Cleanup(runStepCtx)
-		if err != nil {
-			e.logger.Error("cleanup subtask exec env failed", zap.Error(err))
-			// Cleanup is not a critical path of running subtask, so no need to
-			// affect state of subtasks. there might be no subtask to change even
-			// we want to if all subtasks are finished.
-		}
-	}()
+	stepLogger := llog.BeginTask(e.logger.With(
+		zap.String("step", proto.Step2Str(task.Type, task.Step)),
+		zap.Float64("mem-limit-percent", gctuner.GlobalMemoryLimitTuner.GetPercentage()),
+		zap.String("server-mem-limit", memory.ServerMemoryLimitOriginText.Load()),
+		zap.Stringer("resource", resource),
+	), "execute task step")
 
-	for {
-		select {
-		case <-runStepCtx.Done():
-			return runStepCtx.Err()
-		default:
-		}
+	runStepCtx, runStepCancel := context.WithCancelCause(e.ctx)
+	e.stepExec = stepExecutor
+	e.stepCtx = runStepCtx
+	e.stepLogger = stepLogger
 
-		subtask, err := e.taskTable.GetFirstSubtaskInStates(runStepCtx, e.id, task.ID, task.Step,
-			proto.SubtaskStatePending, proto.SubtaskStateRunning)
-		if err != nil {
-			e.logger.Warn("GetFirstSubtaskInStates meets error", zap.Error(err))
-			continue
-		}
-		if subtask == nil {
-			return nil
-		}
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	e.mu.runtimeCancel = runStepCancel
 
-		if subtask.State == proto.SubtaskStateRunning {
-			if !e.IsIdempotent(subtask) {
-				e.logger.Info("subtask in running state and is not idempotent, fail it",
-					zap.Int64("subtask-id", subtask.ID))
-				if err := e.updateSubtaskStateAndErrorImpl(runStepCtx, subtask.ExecID, subtask.ID,
-					proto.SubtaskStateFailed, ErrNonIdempotentSubtask); err != nil {
-					return err
-				}
-				return ErrNonIdempotentSubtask
-			}
-			e.logger.Info("subtask in running state and is idempotent",
+	return nil
+}
+
+func (e *BaseTaskExecutor) cleanStepExecutor() {
+	if err2 := e.stepExec.Cleanup(e.ctx); err2 != nil {
```

Comment:
check `e.stepExec` and `Cleanup` is nil? Not sure it will happen

Reply:
`e.stepExec != nil` is checked at caller